src/pz390.java

```diff
@@ -4671,9 +4720,11 @@ private void ins_B2XX(){
 			ins_setup_ie();
 			break;		
 		case 0xFC: // "B2FC S 7,72 TABORT D2(B2) z15
+			set_psw_check(psw_pic_priv); // RPI 1622 DK Privileged operation exception
 			ins_setup_s();
 			break;		
 		case 0xFF: // 3400 "B2FF" "TRAP4" "S"
+			set_psw_check(psw_pic_spec_op); // RPI 1622 DK Special Operation Exception
 			ins_setup_s();
 			break;
```

Comment:
Principles of Operation (SA22-7832-12), p10-180, states that

'When the PSW control in the trap control block is zero, the operation is as described above except as follows:
- Bit 31 of the current PSW must be zero; otherwise, a special-operation exception is recognized. ..."

Shouldn't these checks be here, before the "set_psw_check(psw_pic_spec_op);" ?

Reply:
The operation is defined as getting the trap control block address from the DUCT. We have no DUCT implemented in z390. We could do, but that would be a major change. For now, changing the operation from a no-op to a S0C2 (at least in problem mode) is an improvement. Further improvements could be made, but I'd rather see a separate story for such a change.

Reply:
At least add a comment. Something like "// See PoOp for TRAP4. Additional checks needed."

Reply:
Also a good suggestion. I'll look up every instruction affected, then add some relevant comments. That should help on the next round of changes. thanks John!